### DISCUSSION TITLE
Goroutine too much

### DISCUSSION THREAD

**Discussion by wind-c** [mochi-mqtt/server#80] [General]
Mochi MQTT creates 3 goroutines for one client，When there are hundreds of thousands or a million of connections, the impact on performance can be significant.
Should be optimized，Please refer to the https://speakerdeck.com/eranyanay/going-infinite-handling-1m-websockets-connections-in-go

**Comment by mochi-co** [comment]
Generally agree but this is likely not possible without a major refactor of the project. There are 3 go routines per client:

- Parent goroutine which sets up and holds the connection of the client itself.
- Reader goroutine which reads data from the connection.
- Writer goroutine which writes data from the connection.

In _theory_ there are ways you could reduce the number of goroutines, but as I can think of, each comes with fairly severe penalties - such as slow clients blocking subsequent connections, or writing blocking reading. 

I would say that hundreds of thousands of connections is currently outside the scope of this project - if you need that type of capacity I would be more comfortable recommending something like NATS or one of the many excellent enterprise MQTT solutions.

That said, always open to optimisation ideas!

**Reply by torntrousers** [reply]
Interesting. Has there been any profiling of Mochi MQTT to see how it its with lots of connections?

I looked through that mentioned slide deck (https://speakerdeck.com/eranyanay/going-infinite-handling-1m-websockets-connections-in-go) which on one slide about their example says "serving a million concurrent connections would consume over 20GB of RAM" - is that really a problem? If i needed to serve a million connections i'd expect to need to use a good sized machine, my laptop has 32GB of RAM, maybe it could serve a million clients with Mochi if I closed all my other running apps?

**Reply by SimonMacIntyre** [reply]
After connecting approximately 3000 clients (each with clean session=true), I noticed our thread count increased to roughly 4 goroutines per client. However, even after all clients disconnected, these threads remained active according to our metrics endpoint. Is this expected behavior, or could it be a memory leak? I'm concerned about the resource utilization if these threads aren't being properly cleaned up.

We waited at least 15 minutes to see if they went back down. Sorry to bump such an old thread!

We have observed that when testing this on a smaller scale, we had +4 goroutines per client. Upon closing them, we only went down -2 goroutine per client, and had 2 leftover.

Edit: came back a couple hours later and it went back down to almost nothing when idle 🤔

**Reply by bddckr** [reply]
I've opened an issue to discuss what Simon's raising above: https://github.com/mochi-mqtt/server/issues/462

**Comment by perbu** [comment]
I don't see a problem with running a few hundred thousand goroutines, fwiw. They're not threads.  Also, starting to optimize before we can measure the impact doesn't sound like a good idea. If we could see a significant performance degradation at some point then that would be a target to work against.
